File: src/usage.ts
/** The audit log. One place writes usage_log; both surfaces use it. */

export interface UsageRow {
  surface: "http" | "mcp";
  route: string;
  targetHost?: string;
  network?: string;
  payer?: string;
  amount?: string;
  txRef?: string;
  paid?: boolean;
  idempotentReplay?: boolean;
  durationMs?: number;
  verdictHash?: string;
  paymentNonce?: string;
}

/**
 * sha256 of a response body. Lets a caller prove a replay returned the same
 * bytes without the log holding the body.
 */
export async function bodyHash(body: unknown): Promise<string> {
  const canonical = JSON.stringify(body, Object.keys(body as object).sort());
  const digest = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(canonical));
  return [...new Uint8Array(digest)].map((b) => b.toString(16).padStart(2, "0")).join("");
}

/** The public audit trail shows the first 6 and last 4 characters of an address. */
export function redactPayer(payer: string): string {
  if (!payer || payer.length <= 10) return payer;
  return `${payer.slice(0, 6)}...${payer.slice(-4)}`;
}

export async function logUsage(db: D1Database, row: UsageRow): Promise<void> {
  await db
    .prepare(
      `INSERT INTO usage_log
         (ts, surface, route, target_host, network, payer, amount, tx_ref,
          paid, idempotent_replay, duration_ms, verdict_hash, payment_nonce)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    )
    .bind(
      new Date().toISOString(),
      row.surface,
      row.route,
      row.targetHost ?? "",
      row.network ?? "",
      row.payer ?? "",
      row.amount ?? "",
      row.txRef ?? "",
      row.paid ? 1 : 0,
      row.idempotentReplay ? 1 : 0,
      row.durationMs ?? 0,
      row.verdictHash ?? "",
      row.paymentNonce ?? "",
    )
    .run();
}

/**
 * Stamp the settlement onto the row the HTTP route already wrote.
 *
 * The route cannot know the tx hash: the SDK middleware settles after the handler
 * returns. So the row goes in unpaid and this runs from the settlement stamp,
 * matching on the payment nonce, which is unique per payment and visible on both
 * sides.
 */
export async function markSettled(
  db: D1Database,
  nonce: string,
  opts: { txRef: string; payer?: string; network?: string },
): Promise<void> {
  if (!nonce) return;
  await db
    .prepare(
      `UPDATE usage_log
          SET paid = 1,
              tx_ref = ?,
              payer = CASE WHEN ? <> '' THEN ? ELSE payer END,
              network = CASE WHEN ? <> '' THEN ? ELSE network END
        WHERE payment_nonce = ? AND paid = 0`,
    )
    .bind(
      opts.txRef,
      opts.payer ?? "",
      opts.payer ?? "",
      opts.network ?? "",
      opts.network ?? "",
      nonce,
    )
    .run();
}

export async function recentUsage(db: D1Database, limit = 25): Promise<unknown[]> {
  const capped = Math.max(1, Math.min(limit, 200));
  const { results } = await db
    .prepare(
      `SELECT ts, surface, route, target_host, network, payer, amount, tx_ref,
              paid, idempotent_replay, duration_ms, verdict_hash
         FROM usage_log ORDER BY ts DESC, id DESC LIMIT ?`,
    )
    .bind(capped)
    .all<Record<string, unknown>>();

  return (results ?? []).map((r) => ({
    ts: r.ts,
    surface: r.surface,
    route: r.route,
    target_host: r.target_host,
    network: r.network,
    payer: redactPayer(String(r.payer ?? "")),
    amount: r.amount,
    tx_ref: r.tx_ref,
    paid: r.paid === 1,
    idempotent_replay: r.idempotent_replay === 1,
    duration_ms: r.duration_ms,
    verdict_hash: r.verdict_hash,
  }));
}

File: app/usage.py
"""The audit log writer/reader. One place writes usage_log, both surfaces use it."""
from __future__ import annotations

import hashlib
import json
from typing import Any

from sqlalchemy import desc, select, update

from .db import SessionLocal
from .models import UsageLog


def body_hash(body: Any) -> str:
    """sha256 of a response body - lets a caller prove a replay returned the
    same bytes without us storing the body in the log."""
    raw = json.dumps(body, sort_keys=True, separators=(",", ":"), default=str)
    return hashlib.sha256(raw.encode()).hexdigest()


def redact_payer(payer: str) -> str:
    """Public audit trail shows the first 6 and last 4 characters of an address."""
    if not payer or len(payer) <= 10:
        return payer
    return f"{payer[:6]}...{payer[-4:]}"


async def log_usage(
    surface: str,
    route: str,
    *,
    target_host: str = "",
    network: str = "",
    payer: str = "",
    amount: str = "",
    tx_ref: str = "",
    paid: bool = False,
    idempotent_replay: bool = False,
    duration_ms: int = 0,
    verdict_hash: str = "",
    payment_nonce: str = "",
) -> None:
    async with SessionLocal() as s:
        s.add(UsageLog(surface=surface, route=route, target_host=target_host,
                       network=network, payer=payer, amount=amount, tx_ref=tx_ref,
                       paid=paid, idempotent_replay=idempotent_replay,
                       duration_ms=duration_ms, verdict_hash=verdict_hash,
                       payment_nonce=payment_nonce))
        await s.commit()


async def mark_settled(nonce: str, *, tx_ref: str, payer: str = "",
                       network: str = "") -> None:
    """Stamp the settlement onto the row the HTTP route already wrote.

    The route cannot know the tx hash: the SDK middleware settles after the
    handler returns. So the row goes in unpaid and this runs from the SDK's
    after-settle hook. Matching on the payment nonce, which is unique per
    payment and visible on both sides.
    """
    if not nonce:
        return
    values = {"paid": True, "tx_ref": tx_ref}
    if payer:
        values["payer"] = payer
    if network:
        values["network"] = network
    async with SessionLocal() as s:
        await s.execute(
            update(UsageLog)
            .where(UsageLog.payment_nonce == nonce, UsageLog.paid.is_(False))
            .values(**values)
        )
        await s.commit()


async def recent_usage(limit: int = 25) -> list[dict[str, Any]]:
    limit = max(1, min(limit, 200))
    async with SessionLocal() as s:
        rows = (
            await s.execute(select(UsageLog).order_by(desc(UsageLog.ts), desc(UsageLog.id)).limit(limit))
        ).scalars().all()
        return [
            {"ts": r.ts.isoformat(), "surface": r.surface, "route": r.route,
             "target_host": r.target_host, "network": r.network,
             "payer": redact_payer(r.payer), "amount": r.amount, "tx_ref": r.tx_ref,
             "paid": r.paid, "idempotent_replay": r.idempotent_replay,
             "duration_ms": r.duration_ms, "verdict_hash": r.verdict_hash}
            for r in rows
        ]
